pkg/pss/evaluate.go

```diff
@@ -0,0 +1,1104 @@
+			if !strings.Contains(check.CheckResult.ForbiddenDetail, containerName) {
+				continue
+			}
+			for _, exclude := range exclude {
+				if !strings.Contains(exclude.RestrictedField, "spec.ephemeralContainers[*]") {
+					continue
+				}
+
+				//	Get values of this container only.
+				// spec.containers[*].securityContext.privileged -> spec.containers[?name=="nginx"].securityContext.privileged
+				newRestrictedField := strings.Replace(restrictedField.path, "*", fmt.Sprintf(`?name=='%s'`, container.Name), 1)
+
+				exempted, err := forbiddenValuesExempted(ctx, pod, check, exclude, newRestrictedField)
+				if err != nil || !exempted {
+					return false, nil
+				}
+				matchedOnce = true
+			}
+			// If container name is in check.Forbidden but isn't exempted by an exclude then pod creation is forbidden
+			if strings.Contains(check.CheckResult.ForbiddenDetail, container.Name) && !matchedOnce {
+				return false, nil
+			}
+		}
+	}
+	return true, nil
+}
+
+func checkPodLevelFields(ctx enginectx.Interface, pod *corev1.Pod, check PSSCheckResult, rule *kyvernov1.PodSecurity, restrictedField restrictedField) (bool, error) {
+	matchedOnce := false
+	for _, exclude := range rule.Exclude {
+		// No exclude for this specific pod-level restrictedField
+		if !strings.Contains(exclude.RestrictedField, restrictedField.path) {
+			continue
+		}
+
+		exempted, err := forbiddenValuesExempted(ctx, pod, check, exclude, exclude.RestrictedField)
+		if err != nil || !exempted {
+			return false, nil
+		}
+		matchedOnce = true
+	}
+	if !matchedOnce {
+		return false, nil
+	}
+	return true, nil
+}
+
+func containsContainerLevelControl(restrictedFields []restrictedField) bool {
+	for _, restrictedField := range restrictedFields {
+		if strings.Contains(restrictedField.path, "ontainers[*]") {
+			return true
+		}
+	}
+	return false
+}
+
+func ExemptProfile(checks []PSSCheckResult, rule *kyvernov1.PodSecurity, pod *corev1.Pod) (bool, error) {
+	ctx := enginectx.NewContext()
+
+	// 1. Iterate over check.RestrictedFields
+	// 2. Check if it's a `container-level` or `pod-level` restrictedField
+	// - `container-level`: container has a disallowed check (container name in check.ForbiddenDetail) && exempted by an exclude rule ? continue : pod creation is forbbiden
+	// - `pod-level`: Exempted by an exclude rule ? good : pod creation is forbbiden
+	for _, check := range checks {
+		for _, restrictedField := range check.RestrictedFields {
+			// Is a container-level restrictedField
+			if strings.Contains(restrictedField.path, "ontainers[*]") {
+				allowed, err := checkContainerLevelFields(ctx, pod, check, rule.Exclude, restrictedField)
+				if err != nil {
+					return false, errors.Wrap(err, err.Error())
+				}
+				if !allowed {
+					return false, nil
+				}
+			} else {
+				// Is a pod-level restrictedField
+				if !strings.Contains(check.CheckResult.ForbiddenDetail, "pod") && containsContainerLevelControl(check.RestrictedFields) {
```

Comment:
`!strings.Contains(check.CheckResult.ForbiddenDetail, "pod")` doesn't seem to be reliable, a container can be named as `nginxpod` for example.

Is the `ForbiddenDetail` starts with "pod" for pod level checks? If so, can we check the prefix?

Reply:
This could be critical for us to facilitate the upstream changes https://github.com/kubernetes/kubernetes/issues/111958.